Python 3.8: asyncio apparently missing to_thread method

Here is another one ...

```
ERROR    aiohttp.server:web_protocol.py:393 Error handling request                                                                      
Traceback (most recent call last):                                                                                                      
  File "/usr/local/lib/python3.8/dist-packages/aiohttp/web_protocol.py", line 422, in _handle_request                                   
    resp = await self._request_handler(request)
  File "/usr/local/lib/python3.8/dist-packages/aiohttp/web_app.py", line 499, in _handle
    resp = await handler(request)                                                                                                       
  File "/home/tobias/caldav/tests/test_caldav.py", line 2196, in xandikos_handler                                                       
    return await mainapp.aiohttp_handler(request, "/")                                                                                  
  File "/usr/local/lib/python3.8/dist-packages/xandikos-0.2.8-py3.8.egg/xandikos/webdav.py", line 2152, in aiohttp_handler
    response = await self._handle_request(request, environ)                                                                             
  File "/usr/local/lib/python3.8/dist-packages/xandikos-0.2.8-py3.8.egg/xandikos/webdav.py", line 2108, in _handle_request              
    return await do.handle(request, environ, self)                                                                                      
  File "/usr/local/lib/python3.8/dist-packages/xandikos-0.2.8-py3.8.egg/xandikos/webdav.py", line 1696, in handle                       
    new_etag = await r.set_body(new_contents, current_etag)                                                                             
  File "/usr/local/lib/python3.8/dist-packages/xandikos-0.2.8-py3.8.egg/xandikos/web.py", line 213, in set_body
    (name, etag) = await asyncio.to_thread(                                                                                             
AttributeError: module 'asyncio' has no attribute 'to_thread'                                                                           
```

--- a/xandikos/web.py
+++ b/xandikos/web.py
@@ -114,6 +114,19 @@ from xandikos.store.git import (
 from xandikos.vcard import VCardFile
 
 
+try:
+    from asyncio import to_thread  # type: ignore
+except ImportError:  # python < 3.8
+    from asyncio import events
+    import contextvars
+
+    async def to_thread(func, *args, **kwargs):  # type: ignore
+        loop = events.get_running_loop()
+        ctx = contextvars.copy_context()
+        func_call = functools.partial(ctx.run, func, *args, **kwargs)
+        return await loop.run_in_executor(None, func_call)
+
+
 WELLKNOWN_DAV_PATHS = {
     caldav.WELLKNOWN_CALDAV_PATH,
     carddav.WELLKNOWN_CARDDAV_PATH,
@@ -199,7 +212,7 @@ class ObjectResource(webdav.Resource):
 
     async def get_file(self) -> File:
         if self._file is None:
-            self._file = await asyncio.to_thread(
+            self._file = await to_thread(
                 self.store.get_file, self.name, self.content_type, self.etag)
             assert self._file is not None
         return self._file
@@ -210,7 +223,7 @@ class ObjectResource(webdav.Resource):
 
     async def set_body(self, data, replace_etag=None):
         try:
-            (name, etag) = await asyncio.to_thread(
+            (name, etag) = await to_thread(
                 self.store.import_one,
                 self.name,
                 self.content_type,

